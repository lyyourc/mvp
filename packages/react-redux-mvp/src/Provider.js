import { Component, Children } from 'react'
import PropTypes from 'prop-types'

export default class Provider extends Component {
  static propTypes = {
    store: PropTypes.object.isRequired,
    children: PropTypes.element.isRequired,
  }

  static childContextTypes = {
    store: PropTypes.object.isRequired,
  }

  constructor(props, context) {
    super(props, context)
  }

  getChildContext = () => {
    return {
      store: this.props.store,
    }
  }

  render() {
    return Children.only(this.props.children)
  }
}
