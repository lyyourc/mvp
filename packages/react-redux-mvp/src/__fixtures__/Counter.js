import React from 'react'
import PropTypes from 'prop-types'

export default class Counter extends React.Component {
  static propTypes = {
    count: PropTypes.number.isRequired,
    handleIncreaseBtnClick: PropTypes.func.isRequired,
    handleDecreaseBtnClick: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props)
  }

  render() {
    const { count, handleIncreaseBtnClick, handleDecreaseBtnClick } = this.props

    return (
      <div>
        <div>{count}</div>
        <button onClick={handleIncreaseBtnClick}>+</button>
        <button onClick={handleDecreaseBtnClick}>-</button>
      </div>
    )
  }
}

Counter.contextTypes = {
  store: PropTypes.object.isRequired,
}
