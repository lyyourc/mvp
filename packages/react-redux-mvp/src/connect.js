import React from 'react'
import PropTypes from 'prop-types'

const noop = () => {}

export default function(mapStateToProps = noop, mapDispatchToProps = noop) {
  return function withConnect(WrappedComponent) {
    return class Connect extends React.Component {
      static contextTypes = {
        store: PropTypes.object.isRequired,
      }

      constructor(props, context) {
        super(props, context)
      }

      componentWillMount() {
        const { store } = this.context

        this.onStateChange()
        this.unsubscribe = store.subscribe(this.onStateChange)
      }

      componentWillUnmount() {
        this.unsubscribe()
      }

      // componentDidUpdate() {
      //   this.onStateChange()
      // }

      onStateChange = () => {
        const { store } = this.context
        const stateProps = mapStateToProps(store.getState())
        const dispatchProps = mapDispatchToProps(store.dispatch)

        this.setState({
          ...stateProps,
          ...dispatchProps,
        })
      }
      
      render() {
        return (
          <WrappedComponent {...this.props} {...this.state} />
        )
      }
    }
  }
}
