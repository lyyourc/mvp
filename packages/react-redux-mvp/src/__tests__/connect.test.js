import React from 'react'
import renderer from 'react-test-renderer'
import connect from '../connect'
import Provider from '../Provider'
import Counter from '../__fixtures__/Counter'
import store from '../__fixtures__/store'

const mapStateToProps = state => ({
  count: state,
})

const mapDispatchToProps = dispatch => ({
  handleIncreaseBtnClick: () => dispatch({ type: 'INCREASE' }),
  handleDecreaseBtnClick: () => dispatch({ type: 'DECREASE' }),
})


test('connnect', () => {
  const CounterContainer = connect(mapStateToProps, mapDispatchToProps)(
    Counter
  )
  const CounterApp = () => {
    return (
      <div>
        <Provider store={store}>
          <CounterContainer />
        </Provider>
      </div>
    )
  }
  const component = renderer.create(CounterApp())
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  const child = component.root.findByType(Counter)
  child.props.handleIncreaseBtnClick()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  child.props.handleDecreaseBtnClick()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
