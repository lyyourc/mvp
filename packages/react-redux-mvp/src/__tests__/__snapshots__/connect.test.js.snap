// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connnect 1`] = `
<div>
  <div>
    <div>
      0
    </div>
    <button
      onClick={[Function]}
    >
      +
    </button>
    <button
      onClick={[Function]}
    >
      -
    </button>
  </div>
</div>
`;

exports[`connnect 2`] = `
<div>
  <div>
    <div>
      1
    </div>
    <button
      onClick={[Function]}
    >
      +
    </button>
    <button
      onClick={[Function]}
    >
      -
    </button>
  </div>
</div>
`;

exports[`connnect 3`] = `
<div>
  <div>
    <div>
      0
    </div>
    <button
      onClick={[Function]}
    >
      +
    </button>
    <button
      onClick={[Function]}
    >
      -
    </button>
  </div>
</div>
`;
